### PR TITLE
feat: add search doc modal

### PR DESCRIPTION
**Pull request by pengx17**

### CODE REVIEW

**Comment by pengx17** [comment]
* **#7157** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7157?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7138** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7136** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7136?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#7135** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7135?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7152** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pengx17 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by pengx17** [comment]
### Merge activity

* **Jun 5, 10:11 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/7136) rebased this pull request after merging its parent, because this pull request is set to merge when ready.
* **Jun 5, 10:22 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/7136) rebased this pull request after merging its parent, because this pull request is set to merge when ready.
* **Jun 6, 2:28 AM EDT**: `pengx17` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).
* **Jun 6, 2:43 AM EDT**: `pengx17`  merged this pull request with the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  constructor(
+    private readonly localState: WorkspaceLocalState,
+    private readonly docsService: DocsService,
+    private readonly workspaceService: WorkspaceService
+  ) {
+    super();
+  }
+  private showCounter = 0;
+  private readonly _mode$ = new LiveData<{
+    counter: number;
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: It seems unnecessary to record this.

If only one quicksearch is displayed at a time, we can hide the older one before showing new one.

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    mode: QuickSearchMode;
+  } | null>(null);
+  private readonly _query$ = new LiveData<string>('');
+  readonly show$ = this._mode$.map(mode => mode !== null);
+
+  private readonly _searchCallbackResult$ = new Subject<SearchCallbackResult>();
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: Maybe we can use a single state livedata?

```ts
state$ = new LiveData<
  {
    mode: QuickSearchMode,
    query: string,
    callback: () => void
  } | null
>(null)
```

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+  toggle() {
+    return this.show$.value ? this.hide() : this.showCMDK();
+  }
+
+  search(query?: string) {
+    const counter = this.show('search-callback');
+    this.setQuery(query || '');
+    return new Promise<SearchCallbackResult | null>(resolve => {
+      const sub = race(
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: Call the callback directly in next `show()`

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  WorkspaceLocalState,
+  WorkspaceService,
+} from '@toeverything/infra';
+import { Entity, LiveData } from '@toeverything/infra';
+import type { Observable } from 'rxjs';
+import { race, skip, Subject } from 'rxjs';
+
+import { resolveLinkToDoc } from '../../navigation';
+
+type QuickSearchMode = 'cmdk' | 'search-callback';
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: I think naming it directly as the type to be searched is more intuitive.

```suggestion
type QuickSearchMode = 'commands' | 'docs';
```

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+  get query$() {
+    return this._query$;
+  }
+
+  setQuery = (query: string) => {
+    this._query$.next(query);
+  };
+
+  addRecentDoc(pageId: string) {
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: Define a standalone service for recent doc?

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+          source: 'link-ref',
+        });
+      }
+    }
+
+    return searchedDocs;
+  }
+
+  private getRecentDocIds() {
+    return this.localState.get(RECENT_PAGES_KEY) || EMPTY_ARRAY;
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]:
```suggestion
    return this.localState.get<string[] | null>(RECENT_PAGES_KEY) || EMPTY_ARRAY;
```

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+  }
+}
+
+declare module '@toeverything/infra' {
+  interface WorkspaceLocalState {
+    get(key: typeof RECENT_PAGES_KEY): string[] | null;
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: You can directly use the generic parameters on WorkspaceLocalState.

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      docId: string;
+      blockId?: string;
+    }
+  | {
+      query: string;
+      action: 'insert';
+    };
+
+// todo: move command registry to entity as well
+export class QuickSearchManager extends Entity {
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: How about name it `QuickSearch` or `QuickSearchPanel`, which can represent an "object" abstracted in our system.
Avoid names like Manager or Worker, as they sound more like a bunch of logic.
This is what I learned in domain-driven design, and I hope to follow such rules.

File: packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts (hunk 7 above)
Comment (pengx17) [reply packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts]: I think overriding `get`, `set` etc will provide better type hints universally even outside of the current module though.

File: packages/frontend/core/src/modules/cmdk/entities/quick-search.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    mode: QuickSearchMode | null = 'commands',
+    opts: {
+      callback?: (res: SearchCallbackResult) => void;
+      query?: string;
+    } = {}
+  ) => {
+    if (mode === null) {
+      this.state$.next(null);
+    } else {
+      this.state$.next({
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cmdk/entities/quick-search.ts:51]:
```suggestion
      if (this.state$.value) {
        this.state$.value.callback(null)
      }
      this.state$.next({
```

Comment (pengx17) [reply packages/frontend/core/src/modules/cmdk/entities/quick-search.ts:51]: Do you mean change `show` to a toggle function? I think could be cases that quick search brought up again even when it is active.

Comment (EYHN) [reply packages/frontend/core/src/modules/cmdk/entities/quick-search.ts:51]: It seems that when opening the next quicksearch, the previous one is not closed, callback(null) is not called anywhere.

Comment (pengx17) [reply packages/frontend/core/src/modules/cmdk/entities/quick-search.ts:51]: Oh I see what you mean.

**Review by EYHN — APPROVED** [review]

**Comment by evanweiguohua** [comment]
Ok

在 2024-06-06 14:23:31，EYHN ***@***.***> 写道：


@EYHN approved this pull request.

**Pull request merged**